include/spotify/json/codec/string.hpp

```diff
@@ -115,17 +115,48 @@ class string_t final {
     detail::fail(context, "\\u must be followed by 4 hex digits");
   }
 
-  static void decode_unicode_escape(decode_context &context, std::string &out) {
+  static void decode_unicode_escape(
+      decode_context &context,
+      std::string &out,
+      unsigned high_surrogate) {
     detail::require_bytes<4>(context, "\\u must be followed by 4 hex digits");
     const auto a = decode_hex_nibble(context, *(context.position++));
```

Comment:
Maybe we can factor out these five lines (decoding the four hex nibbles into a number) into a separate function, and call from within `handle_surrogate_pair` when parsing the low surrogate. I don't like very much that we recurse into the same function to handle the second half of the pair. I think the code will be more elegant and easy to follow if we don't have to check if the `high_surrogate` is non-zero. Something like:

```c++
unsigned decode_hex_number(decode_context &context) {
  detail::require_bytes<4>(context, "\\u must be followed by 4 hex digits");
  const auto a = decode_hex_nibble(context, *(context.position++));
  const auto b = decode_hex_nibble(context, *(context.position++));
  const auto c = decode_hex_nibble(context, *(context.position++));
  const auto d = decode_hex_nibble(context, *(context.position++));
  return unsigned((a << 12) | (b << 8) | (c << 4) | d);
}
```

Reply:
Yes, that seems like a cleaner solution.